src/network/networkd-ndisc.c

```diff
@@ -212,9 +286,29 @@ static int ndisc_router_process_autonomous_prefix(Link *link, sd_ndisc_router *r
         if (r < 0)
                 return log_link_error_errno(link, r, "Failed to get prefix address: %m");
 
-        if (in_addr_is_null(AF_INET6, (const union in_addr_union *) &link->network->ipv6_token) == 0)
-                memcpy(((char *)&address->in_addr.in6) + 8, ((char *)&link->network->ipv6_token) + 8, 8);
-        else {
+        addr = address->in_addr.in6;
+        ORDERED_HASHMAP_FOREACH(j, link->network->ipv6_tokens, i) {
+                if (j->address_generation_type == IPV6_TOKEN_ADDRESS_GENERATION_PREFIXSTABLE
+                    && !memcmp(&j->prefix, &addr, FAMILY_ADDRESS_SIZE(address->family))) {
+                        for (; j->dad_counter < DAD_CONFLICTS_IDGEN_RETRIES_RFC7217; j->dad_counter++) {
+                                r = ipv6_makestableprivate(link, &addr, prefixlen, &j->dad_counter, &address->in_addr.in6);
+                                if (r < 0)
+                                        return log_link_error_errno(link, r, "Could not make stable address: %m");
+
+                                if (!ipv6_interface_identifiers(&address->in_addr.in6)) {
+                                        prefix_stable = true;
+                                        break;
+                                }
+                        }
+                } else if (j->address_generation_type == IPV6_TOKEN_ADDRESS_GENERATION_EUI64) {
+                        if (!memcmp(&j->prefix, &address->in_addr, FAMILY_ADDRESS_SIZE(address->family))) {
+                                memcpy(((char *)&address->in_addr.in6) + 8, ((char *)&j->prefix) + 8, 8);
+                                break;
+                        }
+                }
+        }
+        /* eui64 or fallback if prefixstable do not match */
```

Comment:
This whole chunk (until line 328 in the new numbering) deserves to be moved into its own function.